File: bench/hof/hof.py
import sys, time
from functools import reduce

def sq(x):   return x * x
def pos(x):  return x > 0
def add(a, b): return a + b

def bench(n):
    xs = list(range(n))
    return reduce(add, filter(pos, map(sq, xs)), 0)

N = int(sys.argv[1]) if len(sys.argv) > 1 else 500
for _ in range(100):
    bench(N)
iters = 10000
start = time.monotonic_ns()
for _ in range(iters):
    r = bench(N)
elapsed = time.monotonic_ns() - start
print(f"result:     {r}")
print(f"iterations: {iters}")
print(f"total:      {elapsed / 1e6:.2f}ms")
print(f"per call:   {elapsed // iters}ns")

File: bench/hof/hof.js
function bench(n) {
    const xs = Array.from({length: n}, (_, i) => i);
    return xs.map(x => x * x).filter(x => x > 0).reduce((a, b) => a + b, 0);
}

const N = parseInt(process.argv[2]) || 500;
for (let i = 0; i < 1000; i++) bench(N);
const iters = 10000;
const start = process.hrtime.bigint();
let r;
for (let i = 0; i < iters; i++) r = bench(N);
const elapsed = Number(process.hrtime.bigint() - start);
console.log(`result:     ${r}`);
console.log(`iterations: ${iters}`);
console.log(`total:      ${(elapsed / 1e6).toFixed(2)}ms`);
console.log(`per call:   ${Math.floor(elapsed / iters)}ns`);
